design/dpop.md

```diff
@@ -0,0 +1,308 @@
+
+It is suggested that DPoP should be set up on a per-client basis, with the three options available that should be interpreted as follows:
+
+|**Value** | **Keycloak Server** | **Keycloak Adapters** |
+| --- | --- | --- |
+| Required | OAuth endpoints (where applicable) should require a DPoP header and a DPoP-bound token; should return an error otherwise. | Protected resources should require a DPoP header and a DPoP-bound token, or should return an error otherwise. |
+| Optional | _Token Endpoint_: if DPoP header is present, the returned token should be DPoP-bound and should contain a `cnf` claim with a JWK thumbprint; should be a regular token otherwise  <br/><br/>_UserInfo endpoint_: if the access token is DPoP-bound, a matching DPoP header should be required; no header should be required otherwise. | Protected resources should require a matching DPoP header if the supplied access token is DPoP-bound; should ignore the header otherwise. |
+| Disabled | DPoP headers should be ignored; issued tokens shouldn’t contain JWK thumbprints. | DPoP headers and token JWK thumbprints should be ignored. |
+
+Other DPoP configuration options should include:
+
+| **Option** | **Description** |
+| --- | --- |
+| DPoP Proof Lifetime | Time window in which the respective DPoP proof JWT would be accepted |
+| DPoP Allowed Clock Skew | To accommodate for clock offsets, the server MAY accept DPoP proofs that carry an `iat` time in the reasonably near future (e.g., a few seconds in the future). |
+
+_([8.1. DPoP Proof Replay][4])_
+
+### Endpoints
+
+#### Authorization Endpoint
+
+Binding tokens issued directly from the authorization endpoint has been intentionally considered out of scope for the main DPoP draft, and is defined by a separate draft (see Resources).
+In order to support DPoP in implicit and hybrid flows, authorization endpoint must honor the `dpop` request parameter which would contain a DPoP proof.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint
+
+#### Token Endpoint
+
+With DPoP enabled, token endpoint must understand the DPoP header and perform binding between the keypair presented thereby and the issued token.
+
+As per the spec,
+
+> To request an access token that is bound to a public key using DPoP,
+   the client MUST provide a valid DPoP proof JWT in a "DPoP" header
+   when making an access token request to the authorization server's
+   token endpoint.  This is applicable for all access token requests
+   regardless of grant type (including, for example, the common
+   "authorization\_code" and "refresh\_token" grant types but also
+   extension grants such as the JWT authorization grant \[RFC7523\]).
+
+_([5.5. DPoP Access Token Request][6])_
+
+First and foremost, the `authorization_code` and `refresh_token` grant types must be supported; support for other grant types (ROPC, client credentials, token exchange, extension grants) could be added later.
```

Comment:
Note for refresh tokens: According to the specification `Refresh tokens issued to confidential clients (those having established authentication credentials with the authorization server) are not bound to the DPoP proof public key because they are already sender-constrained with a different existing mechanism.` . So it seems to me that we do not need to bind refresh tokens to DPoP for the confidential clients?

Reply:
Agreed on that; will incorporate into the prototype implementation.